scripts/build-scenario.ts

```diff
@@ -0,0 +1,73 @@
+// Build a scenario with given ID from CLI args
+
+// Usage: `yarn scenario [ID]`
+// ID = name of scenario script OR 'default' if no ID provided
+
+import { ScenarioBuilder, Scenario } from './content-utils'
+import { promises } from 'fs'
+import { GameWorld } from '../src/game/ContentTypes'
+const { writeFile, mkdir } = promises
+
+const id = process.argv.length >= 3 ? process.argv[2] : 'default'
+
+async function buildScenario(id: string) {
+    try {
+        const builder: ScenarioBuilder = (
+            await import(`./scenarios/${id}/index.ts`)
+        ).builder
+
+        const scenario = builder.run()
+
+        console.log(JSON.stringify(scenario, null, 2))
+        console.log(`✅ Sucessfully built scenario "${id}"`)
+
+        return scenario
+    } catch (e) {
+        throw new Error(`Error: Failed to build scenario "${id}" - ${e}`)
+    }
+}
+
+async function exportScenario(id: string, scenario: Scenario) {
+    // TODO: export scenario to JSON
+    // Build into separate files and store in a output directory.
+    // Create files in a directory, or write to existing if they already have been created.
+
+    type Parts = {
+        [K in keyof GameWorld]: string
+    }
+
+    const scenarioParts: Parts = {
+        cards: 'cards',
+        eventCards: 'event-cards',
+        events: 'events',
+        defaultState: 'default-state',
+        stats: 'stats',
+    }
+
+    // Create the necessary directories if they are missing
+    const outputDir = `${__dirname}/build/${id}`
+    await mkdir(outputDir, { recursive: true })
+
+    await Promise.all(
+        Object.entries(scenarioParts).map(([part, fileName]) =>
+            writeFile(
+                `${outputDir}/${fileName}.json`,
+                JSON.stringify(scenario[part as keyof GameWorld], null, 4),
+            ),
+        ),
+    )
+
+    // TODO: improve the export, verify that it works correcly by diffing JSON output with default scenario
+    // TODO: Export directly to the /src/public/scenarios directory to make it playable in game
+    // TODO: Test export in the game and see if it's playable
+    // TODO: simplify prop references to make output cleaner, since data might be edited manually in other editors - and that the generated prop references will change on every build.
+
+    // IDEA: Add support for building all images used and saving them to a directory
+    // save with format `pexels-[ID].png` or `unsplash-[ID].png`
```

Comment:
Do you have a good reason to even consider keeping the reference to where the images were source from? I'd suggest using a similar system as propRef and cardRef, perhaps imageRef, to generate unique image names. If you want to keep a reference to the sources I'd suggest generating a separate json mapping the new image URLS names to the old URLs. This way we won't have to consider the source of the image when downloading and generating the stand alone content.

Reply:
Yes, I think there is value in naming the exported images `unsplahsh- [ID].png` and similar to always know where to find the original image, even in exported format.

This is a way of giving credit to the photographers as we could link back to their photos.

Reply:
This will be solved in https://github.com/Greenheart/swipeforfuture.com/issues/110